File: sdk/go/process.go
package axernsdk

import (
	"context"
	"errors"
	"io"
	"iter"
	"sync"
	"time"

	"github.com/cofy-x/axern/sdk/go/internal/nodeclient"
)

// ProcessOptions configures an attached sandbox process.
type ProcessOptions struct {
	Env          map[string]string
	Cwd          string
	Timeout      time.Duration
	User         string
	TTY          bool
	ManagedProxy *ManagedProxyOptions
}

// ProcessEventKind identifies a process stream event.
type ProcessEventKind string

const (
	ProcessEventStdout ProcessEventKind = "stdout"
	ProcessEventStderr ProcessEventKind = "stderr"
	ProcessEventExit   ProcessEventKind = "exit"
)

// ProcessEvent is a stdout, stderr, or exit event from an attached process.
type ProcessEvent struct {
	Kind               ProcessEventKind
	Data               []byte
	ExitCode           int32
	Message            string
	ManagedProxyReport *ManagedProxyReport
}

// ProcessResult is the exit status for a sandbox process.
type ProcessResult struct {
	ExitCode           int32
	Message            string
	ManagedProxyReport *ManagedProxyReport
}

// ProcessOutput contains collected stdout, stderr, and exit status.
type ProcessOutput struct {
	ProcessResult
	Stdout []byte
	Stderr []byte
}

// SandboxProcess is an attached sandbox process with stream controls.
type SandboxProcess struct {
	allocationID string
	process      sandboxProcessClient
	sandbox      *Sandbox
	mu           sync.Mutex
	exit         *ProcessResult
	closeOnce    sync.Once
	closeErr     error
}

type sandboxProcessClient interface {
	Write([]byte) error
	CloseStdin() error
	Resize(uint32, uint32) error
	Signal(string) error
	Recv() (nodeclient.ProcessEvent, error)
	Close() error
}

// Process starts an attached process in the sandbox.
func (s *Sandbox) Process(ctx context.Context, command any, options ProcessOptions) (*SandboxProcess, error) {
	node, err := s.nodeClient()
	if err != nil {
		return nil, err
	}
	process, err := node.Process(ctx, command, options)
	if err != nil {
		return nil, err
	}
	process.sandbox = s
	s.registerProcess(process)
	return process, nil
}

// Process starts an attached process in the allocation.
func (n *NodeSandboxClient) Process(ctx context.Context, command any, options ProcessOptions) (*SandboxProcess, error) {
	if err := n.validate(); err != nil {
		return nil, err
	}
	if err := validateProcessOptions(options); err != nil {
		return nil, err
	}
	argv, err := normalizeCommand(command)
	if err != nil {
		return nil, err
	}
	process, err := n.rpcClient().Process(ctx, argv, nodeclient.Options{
		Env:          options.Env,
		Cwd:          options.Cwd,
		Timeout:      options.Timeout,
		User:         options.User,
		TTY:          options.TTY,
		ManagedProxy: nodeManagedProxyOptions(options.ManagedProxy),
	})
	if err != nil {
		return nil, mapRPCError(err, "sandbox process", n.allocationID)
	}
	return &SandboxProcess{allocationID: n.allocationID, process: process}, nil
}

// Write sends bytes to the process stdin stream.
func (p *SandboxProcess) Write(data []byte) error {
	if p == nil || p.process == nil {
		return ErrProcessClosed
	}
	return mapProcessError(p.process.Write(data), "sandbox process write", p.allocationID)
}

// WriteString sends a string to the process stdin stream.
func (p *SandboxProcess) WriteString(data string) error {
	return p.Write([]byte(data))
}

// CloseStdin closes the process stdin stream.
func (p *SandboxProcess) CloseStdin() error {
	if p == nil || p.process == nil {
		return ErrProcessClosed
	}
	return mapProcessError(p.process.CloseStdin(), "sandbox process close stdin", p.allocationID)
}

// Resize sends a terminal resize event for TTY processes.
func (p *SandboxProcess) Resize(cols, rows uint32) error {
	if p == nil || p.process == nil {
		return ErrProcessClosed
	}
	if cols == 0 {
		return validationError("cols", "must be greater than zero")
	}
	if rows == 0 {
		return validationError("rows", "must be greater than zero")
	}
	return mapProcessError(p.process.Resize(cols, rows), "sandbox process resize", p.allocationID)
}

// Signal sends a named signal to the process.
func (p *SandboxProcess) Signal(signal string) error {
	if p == nil || p.process == nil {
		return ErrProcessClosed
	}
	if signal == "" {
		return requiredError("signal")
	}
	return mapProcessError(p.process.Signal(signal), "sandbox process signal", p.allocationID)
}

// Terminate sends SIGTERM to the process.
func (p *SandboxProcess) Terminate() error {
	return p.Signal("TERM")
}

// Kill sends SIGKILL to the process.
func (p *SandboxProcess) Kill() error {
	return p.Signal("KILL")
}

// Recv receives the next process event.
func (p *SandboxProcess) Recv() (ProcessEvent, error) {
	if p == nil || p.process == nil {
		return ProcessEvent{}, ErrProcessClosed
	}
	event, err := p.process.Recv()
	if err != nil {
		if errors.Is(err, io.EOF) {
			return ProcessEvent{}, err
		}
		return ProcessEvent{}, mapProcessError(err, "sandbox process recv", p.allocationID)
	}
	sdkEvent := ProcessEvent{
		Kind:               ProcessEventKind(event.Kind),
		Data:               event.Data,
		ExitCode:           event.ExitCode,
		Message:            event.Message,
		ManagedProxyReport: sdkManagedProxyReport(event.ManagedProxyReport),
	}
	if sdkEvent.Kind == ProcessEventExit {
		p.rememberExit(ProcessResult{
			ExitCode:           sdkEvent.ExitCode,
			Message:            sdkEvent.Message,
			ManagedProxyReport: sdkEvent.ManagedProxyReport,
		})
		_ = p.closeProcess()
	}
	return sdkEvent, nil
}

// Events returns an iterator over process events until exit or error.
func (p *SandboxProcess) Events() iter.Seq2[ProcessEvent, error] {
	return func(yield func(ProcessEvent, error) bool) {
		for {
			event, err := p.Recv()
			if err != nil {
				if !errors.Is(err, io.EOF) {
					yield(ProcessEvent{}, err)
				}
				return
			}
			if !yield(event, nil) || event.Kind == ProcessEventExit {
				return
			}
		}
	}
}

// Wait drains events until the process exit status is observed.
func (p *SandboxProcess) Wait() (ProcessResult, error) {
	if p == nil || p.process == nil {
		return ProcessResult{}, ErrProcessClosed
	}
	p.mu.Lock()
	if p.exit != nil {
		result := *p.exit
		p.mu.Unlock()
		return result, nil
	}
	p.mu.Unlock()

	for {
		event, err := p.Recv()
		if err != nil {
			if errors.Is(err, io.EOF) {
				_ = p.closeProcess()
				return ProcessResult{}, ErrProcessExitMissing
			}
			return ProcessResult{}, err
		}
		if event.Kind == ProcessEventExit {
			return ProcessResult{
				ExitCode:           event.ExitCode,
				Message:            event.Message,
				ManagedProxyReport: event.ManagedProxyReport,
			}, nil
		}
	}
}

// Output collects stdout, stderr, and exit status.
func (p *SandboxProcess) Output() (ProcessOutput, error) {
	var output ProcessOutput
	for event, err := range p.Events() {
		if err != nil {
			return output, err
		}
		switch event.Kind {
		case ProcessEventStdout:
			output.Stdout = append(output.Stdout, event.Data...)
		case ProcessEventStderr:
			output.Stderr = append(output.Stderr, event.Data...)
		case ProcessEventExit:
			output.ProcessResult = ProcessResult{
				ExitCode:           event.ExitCode,
				Message:            event.Message,
				ManagedProxyReport: event.ManagedProxyReport,
			}
			return output, nil
		}
	}
	return output, ErrProcessExitMissing
}

// Close closes the attached process stream.
func (p *SandboxProcess) Close() error {
	if p == nil || p.process == nil {
		return nil
	}
	return p.closeProcess()
}

func (p *SandboxProcess) rememberExit(result ProcessResult) {
	p.mu.Lock()
	if p.exit == nil {
		p.exit = &result
	}
	p.mu.Unlock()
}

func (p *SandboxProcess) closeProcess() error {
	p.closeOnce.Do(func() {
		p.closeErr = p.process.Close()
		if p.sandbox != nil {
			p.sandbox.unregisterProcess(p)
		}
	})
	return p.closeErr
}

func mapProcessError(err error, operation, allocationID string) error {
	if errors.Is(err, nodeclient.ErrProcessClosed) {
		return ErrProcessClosed
	}
	return mapRPCError(err, operation, allocationID)
}

func validateProcessOptions(options ProcessOptions) error {
	if options.Timeout < 0 {
		return positiveDurationError("timeout")
	}
	return nil
}

File: sdk/typescript/src/node/process.ts
/**
 * @license
 * Copyright 2026 cofy-x
 * SPDX-License-Identifier: Apache-2.0
 */

import type * as grpc from "@grpc/grpc-js";

import { mapRpcError } from "../errors/index.js";
import type { ProcessEvent, ProcessResult } from "../types.js";

export interface SandboxProcessOptions {
  allocationId: string;
  call: grpc.ClientDuplexStream<Record<string, unknown>, Record<string, unknown>>;
  closeClient: () => void;
}

export class SandboxProcess {
  readonly allocationId: string;

  private readonly call: grpc.ClientDuplexStream<Record<string, unknown>, Record<string, unknown>>;
  private readonly closeClient: () => void;
  private readonly queue: ProcessEvent[] = [];
  private readonly waiters: Array<(value: IteratorResult<ProcessEvent>) => void> = [];
  private error: unknown;
  private ended = false;
  private exit?: ProcessResult;
  private paused = false;
  private closed = false;

  private static readonly maxQueuedEvents = 64;
  private static readonly resumeQueuedEvents = 32;

  constructor(options: SandboxProcessOptions) {
    this.allocationId = options.allocationId;
    this.call = options.call;
    this.closeClient = options.closeClient;

    this.call.on("data", (message: Record<string, unknown>) => this.enqueue(processEvent(message)));
    this.call.on("error", (error: unknown) => {
      this.error = error;
      this.finish();
    });
    this.call.on("end", () => this.finish());
  }

  write(data: Buffer | Uint8Array | string): Promise<void> {
    return this.writeRequest({ stdin: Buffer.isBuffer(data) ? data : Buffer.from(data) });
  }

  closeStdin(): Promise<void> {
    return this.writeRequest({ close_stdin: true });
  }

  resize(cols: number, rows: number): Promise<void> {
    return this.writeRequest({ resize: { cols, rows } });
  }

  signal(signal: string): Promise<void> {
    return this.writeRequest({ signal: { signal } });
  }

  terminate(): Promise<void> {
    return this.signal("TERM");
  }

  kill(): Promise<void> {
    return this.signal("KILL");
  }

  async waitReady(): Promise<void> {
    const next = await this.nextEvent();
    if (next.done === true) {
      if (this.error !== undefined) {
        throw this.error;
      }
      throw new Error("sandbox process stream ended before ready");
    }
    if (next.value.kind !== "ready") {
      this.queue.unshift(next.value);
    }
  }

  async wait(): Promise<ProcessResult> {
    if (this.exit !== undefined) {
      return this.exit;
    }
    for await (const event of this.events()) {
      if (event.kind === "exit") {
        return { exitCode: event.exitCode, message: event.message };
      }
    }
    if (this.error !== undefined) {
      throw mapRpcError(this.error, "sandbox process", this.allocationId);
    }
    throw new Error("sandbox process stream ended before exit");
  }

  async close(): Promise<void> {
    if (this.closed) return;
    if (!this.ended && this.exit === undefined) {
      await this.terminate().catch(() => undefined);
    }
    this.closed = true;
    this.call.end();
    this.call.cancel();
    this.closeClient();
  }

  async *events(): AsyncIterable<ProcessEvent> {
    while (true) {
      const next = await this.nextEvent();
      if (next.done === true) {
        if (this.error !== undefined) {
          throw mapRpcError(this.error, "sandbox process", this.allocationId);
        }
        return;
      }
      yield next.value;
      if (next.value.kind === "exit") {
        return;
      }
    }
  }

  private nextEvent(): Promise<IteratorResult<ProcessEvent>> {
    const event = this.queue.shift();
    if (event !== undefined) {
      if (this.paused && this.queue.length <= SandboxProcess.resumeQueuedEvents) {
        this.paused = false;
        this.call.resume();
      }
      return Promise.resolve({ done: false, value: event });
    }
    if (this.ended) {
      return Promise.resolve({ done: true, value: undefined });
    }
    return new Promise((resolve) => this.waiters.push(resolve));
  }

  private enqueue(event: ProcessEvent): void {
    if (event.kind === "exit") {
      this.exit = { exitCode: event.exitCode, message: event.message };
      this.call.end();
      this.closeClient();
    }
    const waiter = this.waiters.shift();
    if (waiter !== undefined) {
      waiter({ done: false, value: event });
    } else {
      if (this.queue.length >= SandboxProcess.maxQueuedEvents) {
        this.error = new Error("sandbox process consumer is too slow; event queue limit exceeded");
        this.call.cancel();
        this.finish();
        return;
      }
      this.queue.push(event);
      if (!this.paused && this.queue.length >= SandboxProcess.maxQueuedEvents) {
        this.paused = true;
        this.call.pause();
      }
    }
  }

  private writeRequest(request: Record<string, unknown>): Promise<void> {
    if (this.closed || this.ended) {
      return Promise.reject(new Error("sandbox process stream is closed"));
    }
    return new Promise((resolve, reject) => {
      this.call.write(request, (error?: Error | null) => {
        if (error !== undefined && error !== null) {
          reject(mapRpcError(error, "sandbox process write", this.allocationId));
          return;
        }
        resolve();
      });
    });
  }

  private finish(): void {
    if (this.ended) {
      return;
    }
    this.ended = true;
    this.closeClient();
    for (const waiter of this.waiters.splice(0)) {
      waiter({ done: true, value: undefined });
    }
  }
}

function processEvent(message: Record<string, unknown>): ProcessEvent {
  if ("ready" in message) {
    return { kind: "ready" };
  }
  if ("stdout" in message) {
    return { kind: "stdout", data: Buffer.from((message.stdout as Buffer | Uint8Array | undefined) ?? []) };
  }
  if ("stderr" in message) {
    return { kind: "stderr", data: Buffer.from((message.stderr as Buffer | Uint8Array | undefined) ?? []) };
  }
  if ("exit" in message) {
    const exit = message.exit as Record<string, unknown> | undefined;
    return {
      kind: "exit",
      exitCode: Number(exit?.exit_code ?? 0),
      message: String(exit?.message ?? ""),
    };
  }
  return { kind: "ready" };
}
